File: common/src/main/java/cc/turtl/chiselmon/system/dashboard/api/endpoints/DimensionsHandler.java
package cc.turtl.chiselmon.system.dashboard.api.endpoints;

import cc.turtl.chiselmon.system.dashboard.api.ApiHandler;
import cc.turtl.chiselmon.system.tracker.EncounterDatabase;
import com.sun.net.httpserver.HttpExchange;

import java.io.IOException;
import java.util.List;

public class DimensionsHandler extends ApiHandler {

    private record DimensionEntry(String dimension, long count) {
    }

    private record DimensionsResponse(List<DimensionEntry> dimensions) {
    }

    public DimensionsHandler(EncounterDatabase db) {
        super(db);
    }

    @Override
    public void handle(HttpExchange exchange) throws IOException {
        handleRequest(exchange, (timeRange, params) -> {
            List<DimensionEntry> dimensions = query("encounters")
                    .timeRange(timeRange)
                    .select("dimension, COUNT(*) as cnt")
                    .groupBy("dimension")
                    .orderBy("cnt DESC")
                    .fetchList(rs -> new DimensionEntry(rs.getString("dimension"), rs.getLong("cnt")));

            return new DimensionsResponse(dimensions);
        });
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/system/dashboard/api/endpoints/DimensionsHandler.kt
package cc.turtl.chiselmon.client.system.dashboard.api.endpoints

import cc.turtl.chiselmon.client.system.dashboard.api.ApiHandler
import cc.turtl.chiselmon.client.system.tracker.EncounterDatabase
import com.sun.net.httpserver.HttpExchange

class DimensionsHandler(db: EncounterDatabase) : ApiHandler(db) {

    private data class DimensionEntry(val dimension: String, val count: Long)
    private data class DimensionsResponse(val dimensions: List<DimensionEntry>)

    override fun handle(exchange: HttpExchange) {
        handleRequest(exchange) { timeRange, _ ->
            val dimensions = query("encounters")
                .timeRange(timeRange)
                .select("dimension, COUNT(*) as cnt")
                .groupBy("dimension")
                .orderBy("cnt DESC")
                .fetchList { rs -> DimensionEntry(rs.getString("dimension"), rs.getLong("cnt")) }
            DimensionsResponse(dimensions)
        }
    }
}
